### DISCUSSION TITLE
Templating Tools and GitOps

### DISCUSSION THREAD

**Discussion by williamcaban** [open-gitops/project#31] [Ideas]
# Templating Tools and GitOps

There are multiple templating techniques used by various tools. Some are easy to see how they work with the GitOps principles, while for others is not so clear. This is a discussion about best practices or patterns to follow when using templating tools on different scenarios to make sure the GitOps principles are been followed.

For a repositories `repo1`, `repo2`, ... , `repoN`
For a managed system `system1`, `system2`, ... , `systemN`

## Case 1

- Templating tool reads a template from `repo1`, and read data from same or different repository
- During continuous reconciliation the templating tool combines template & data in memory, and apply desired state to `system1`

## Case 2

- Templating tool reads a template from `repo1`, and read data from same or different repository
- Templating tool combines template & data, and writes results to `repo3`

Under which setup can this scenario be considered following GitOps principles?

## Case 3

- Templating tool has its own storage backend with a template and reads data from the backend
- Templating tool combines template & data, and writes results to `repo1`

3.1 storage backend not versioned
3.2 storage backend versioned

## Case 4

- Templating tool has its own storage backend with a template and reads data from the backend
- Templating tool combines template & data, and writes result to the backend
- A reconciliation tool read from the backend and applies reconciliation to `system1`

4.1 storage backend not versioned
4.2 storage backend versioned

## Case 5

- Templating tool has its own storage backend with a template and reads data from the backend
- Templating tool combines template & data, and applies the resuts to `system1`
- Templating tool does continue reconciliation of `system1`

5.1 storage backend not versioned
5.2 storage backend versioned

## Case 6

- Templating tool requires client plugin

Under which setup can this scenario be considered following GitOps principles?

**Comment by lloydchang** [comment]
It depends on the tool.

I tend to use Kustomize as the standard, because Kustomize is built into Kubernetes, and CNCF promotes Kubernetes.

If we are to spend time in GitOps Working Group on templating tools, then we should start with Kustomize.

Are the tools you're thinking of different than Kustomize?  If so, what are they?  Are they vendor-neutral tools?

Thank you for the discussion.

**Reply by williamcaban** [reply]
It is about the different patterns followed when using templating tools, not about the tools themselves.

**Reply by lloydchang** [reply]
@williamcaban Pardon my ignorance for asking... and sorry for my lack of understanding of the primary objective of this discussion.

Defining the primary objective:
• What problem is this discussion trying to solve?
• Is it the lack of clarity of unnamed tools `a` to `z` with regards to GitOps?
• Is this discussion about best practices and patterns trending toward the scenario: A solution that's looking for a problem to solve?
• Perhaps the problem is already solved?
• Are we discussing generic tools in the abstract?
• Perhaps something else all together?

I believe we shouldn't spend time discussing best practices or patterns if there isn't a real problem with a named tool that needs solving.

I would rather have the conversation about...

• For `tool1`
• For a repositories `repo1`, `repo2`, ... , `repoN`
• For a managed system `system1`, `system2`, ... , `systemN`

Instead of generically:
• For unnamed tools `a` to `z`
• For a repositories `repo1`, `repo2`, ... , `repoN`
• For a managed system `system1`, `system2`, ... , `systemN`

Knowing what that `tool1` is impacts the use cases presented, and the best practices or patterns.

Each tool has features and limitations that can impact the different best practices and patterns that follow.

The reason I suggest starting with Kustomize is because Kustomize is a part of Kubernetes.

I may need a refresher in the following:

https://github.com/kubernetes/community/blob/master/contributors/design-proposals/architecture/declarative-application-management.md
explains Kustomize's design and architecture.

There are examples at
https://github.com/kubernetes-sigs/kustomize/tree/master/examples
with more resources at https://kustomize.io/

Flux CD has a Kustomizer controller
• https://github.com/fluxcd/kustomize-controller

Argo CD interacts with Kustomize
• https://argo-cd.readthedocs.io/en/stable/user-guide/kustomize/

_**If we're using a different tool than Kustomize, then I believe the answers about the different best practices or patterns will vary.**_

For example: Flux CD and Argo CD interact with Helm differently than their interactions with Kustomize.
• https://github.com/fluxcd/helm-controller
• https://argo-cd.readthedocs.io/en/stable/user-guide/helm/

This is one area in which I agree with @roberthstrand's commentary that:
> "It’s the tools that make GitOps, to be honest."
> https://youtu.be/jGYH5oxKv3o?t=5879s

Thank you @williamcaban and @roberthstrand for your time.
